Add some CF support for lon and lat

stefraynaud opened the pull request:
This PR adds some support for CF conventions to search for longitude and latitude in Datasets, and rename them to 'lon' and 'lat'. Bounds data arrays are also renamed if found.

A module named 'cf' was created for that and unitests were added.

stefraynaud committed:
Add some CF support for lon and lat

stefraynaud committed:
Add name restoration to CF renaming for outputs

stefraynaud commented:
With the first commit, coordinates and bounds are automatically identified and renamed, if they respect CF conventions which are based on name, units and standard_name.

With the second commit, output coordinate names are automatically restored.

JiaweiZhuang commented:
@stefraynaud Thanks for the PR! I had a long backlog and didn't get time to look at this -- sorry for the delay.

I agree that some degree of decoding feature is useful. On the other hand I am also worried about putting complex logic into xesmf, especially tricky edges cases like: which variable actually got used if the dataset contains all three variables `'lat_b'`, `'lat_bnds'` `'latitude_bnds'`. 

To that end I summarized a meta-issue at #74. Would you be able to build the CF-convention support on top of the proposed `xesmf.config.set(grid_name_dict=...)`? That would be much easier to maintain and extend.

Also, will [`xarray.decode_cf`](http://xarray.pydata.org/en/stable/generated/xarray.decode_cf.html) be useful here?

JiaweiZhuang commented on xesmf/cf.py:
```diff
@@ -0,0 +1,158 @@
+CF_SPECS = {
+    'lon': {
+        'name': ['lon', 'longitude'],
+        'units': ['degrees_east', 'degree_east', 'degree_E', 'degrees_E',
```
Checking the units and other attributes seems a bit too heavy for xesmf? A "CF-checker" is better for a standalone package or pushed to upstream like [xarray.decode_cf](http://xarray.pydata.org/en/stable/generated/xarray.decode_cf.html)

stefraynaud replied:
I agree with you that a CF checker for xarray is globally missing. 
However, xarray is intended to remain of general purpose, and not to introduce notions of geographic and physical spatialisation.

As for the fact that these attributes are checked is the strict application of the CF conventions for [longitude ](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#longitude-coordinate) and latitude. 
This prevent focusing on names.

JiaweiZhuang commented on xesmf/cf.py:
```diff
@@ -0,0 +1,158 @@
+    ds: xarray.DataArray or xarray.Dataset
+    '''
+    # Suffixes for bounds
+    bsuffixes = ('_b', '_bounds', '_bnds')
```
Here assumes a priority of '_b' > '_bounds' > '_bnds'. If more than one of them exist in the dataset, could it lead to a name conflict error?

stefraynaud replied:
This is the main caveat of this PR because it rely on names and not on attributes.

Don't blame me : a solution is to push forward the application of CF conventions and to use the ``bounds`` attribute as explained [here](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#cell-boundaries). This makes unique link between coordinates and bounds.

JiaweiZhuang commented:
Also note renaming the coordinate name alone does not solve CF-convention -- CF uses a boundary shape of `(n_lat, n_lon, 4)` while ESMPy expects `(n_lat+1, n_lon+1)` https://github.com/JiaweiZhuang/xESMF/issues/14#issuecomment-369686779.

stefraynaud commented:
> 
> 
> Also note that renaming the coordinate name alone does not solve CF-convention -- CF uses a boundary shape of `(n_lat, n_lon, 4)` while ESMPy expects `(n_lat+1, n_lon+1)` [#14 (comment)](https://github.com/JiaweiZhuang/xESMF/issues/14#issuecomment-369686779).

You are right, but this is beyond the scope of this PR which is focused on finding variables in datasets.

By the way, a conversion from the ``(n_lat, n_lon, 4)`` form to the ``(n_lat+1, n_lon+1)`` it straigthforward, but must raise a warning (my two cents).

stefraynaud committed:
Add CF support to search for bounds

It first look ate the "bounds" attribute, then search
for names using usual prefixes.

huard commented on xesmf/cf.py:
```diff
@@ -0,0 +1,203 @@
+    }
+
+
+def get_coord_name_from_specs(ds, specs):
```
This is certainly a more extensive search than what #74 provides. The question is do we want this ? Also, from a user perspective, all the different ways coordinates can be identified might create some confusion.

stefraynaud replied:
In fact, it is the strict application of the CF conventions, which must be considered to be the reference for formatting dataset, and which are increasingly being followed.

huard replied:
See https://github.com/JiaweiZhuang/xESMF/pull/94#issuecomment-618699019

stefraynaud replied:
Yep. Checking with units first since they are mandatory. And then standard_name, just in case.

huard commented on xesmf/cf.py:
```diff
@@ -0,0 +1,203 @@
+
+
+def get_bounds_name_from_coord(ds, coord_name,
+                               suffixes=['_b', '_bnds', '_bounds']):
```
I think this could get really confusing, because this function could either return netCDF `bounds` or the grid corners (`_b`). I suggest the code clearly distinguishes bounds from corners.

stefraynaud replied:
Yep probably. I made no difference in this PR regarding the bounds/corners problem.

huard commented on xesmf/cf.py:
```diff
@@ -0,0 +1,203 @@
+            return bounds_name
+        warnings.warn('invalid bounds name: ' + bounds_name)
+
+    # From suffixed names
```
I'd remove the identification by suffix.

huard commented on xesmf/cf.py:
```diff
@@ -0,0 +1,203 @@
+    lon_name = get_lon_name(ds)
+    if lon_name is not None:
+
+        # Search for bounds and rename
```
We should not have to rename coordinates.

stefraynaud replied:
In fact, renaming is for internal use only, just to make the code clearer. It is possible to not rename :)

huard reviewed (commented):
I think we should try to avoid renaming coordinates. 
I believe the search for coordinates might be too complex to be clearly communicated to users. I can imagine questions such as "why did it work on this file and not this one ?"
We should strive to reduce the potential confusion between CF bounds and grid corners.
